Added an optional timeout to the Mailbox.take method

bsutton opened the pull request:
This required changes to the  underlying Mutex.runLocked and ConditionalVariable.wait methods.


---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

The changes are backwards compatible simply adding an optional 'timeout' argument to each of the noted methods.

The changes provide greater flexibility in how the methods are used and will also aid in debugging dead locks (which is what prompted me to do the work).

bsutton committed:
Added an optional timeout to the Mailbox.take method as well as the underlying Mutex.runLocked and ConditionalVariable.wait methods.

bsutton committed:
updated the change log.

bsutton committed:
Made the TimeoutException messages more consistent.
Fixed a lint for Timeout Test

bsutton committed:
ignored .failed_tracker

bsutton committed:
Released 0.4.0.

bsutton committed:
Released 0.4.0.

bsutton commented:
I've released a temp package to pub.dev if anyone needs this functionality urgently
https://pub.dev/packages/native_synchronization_temp

mraleph commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+# 0.4.0
```
Please merge with `main`.

bsutton replied:
done.

mraleph commented on lib/mailbox.dart:
```diff
@@ -7,8 +7,8 @@ import 'dart:typed_data';
 
 import 'package:ffi/ffi.dart';
 
-import 'package:native_synchronization/primitives.dart';
```
nit: unless there is a strong reason please don't change formatting of the code which you are not changing - that makes review simpler.

bsutton replied:
It should all be formatted with dart format.

bsutton replied:
I will check that dart format has been run before I submit a pr update.

mraleph commented on lib/mailbox.dart:
```diff
@@ -48,9 +48,11 @@ class Mailbox {
-    calloc.free(mailbox.ref.buffer);
-    calloc.free(mailbox);
+  static final finalizer = Finalizer((mailbox) {
+    final p = mailbox! as Pointer<_MailboxRepr>;
```
Why change this?

bsutton replied:
because I dev using strong type lint options and the linter was complaining, so I fixed it.

mraleph replied:
What was it complaining about? The change makes the code less typed: `finalizer` had static type `Finalizer<Pointer<_MailboxRepr>>` after your change this type is lost and you are resorting to runtime casting for some reason.

Please revert this change.

mraleph replied:
This is still not resolved

mraleph commented on lib/mailbox.dart:
```diff
@@ -87,7 +90,44 @@ class Mailbox {
+    return _mutex.runLocked(
+      timeout: timeout,
+      () {
+        timeout = timeout - (DateTime.now().difference(start));
```
This code needs to be inside the loop because you need to update timeout if you woke up spuriously. 

I suggest to write this approximately like this:

```dart
final timeoutBy = timeout != null ? DateTime.now().add(timeout) : null;

return _mutex.runLocked(timeout: timeout, () {
  while (_mailbox.ref.state != ...) {
    _condVar.wait(_mutex, timeout: timeoutBy?.difference(DateTime.now()));
  }
});
```

I also think that timed and non-timed version can basically be the same. (just pass null timeout to `runLocked` and `wait`).

bsutton replied:
I've modified the code to recalc the remaining time within the loop.

I considered using same method but decided the simplicity of two methods would aid long term maintenance.

mraleph commented on lib/posix.dart:
```diff
@@ -31,8 +35,33 @@ class _PosixMutex extends Mutex {
+    final timespec =
+        malloc.allocate<pthread_timespec_t>(sizeOf<pthread_timespec_t>());
+
+    /// calculate the absolute timeout in seconds
```
nit: `// Calculate ...`

nit: `abstimeout` is better called something like `wakeupTime` or something like that.

Also seems strange to just discard milliseconds given that underlying OS supports timeouts with larger precision.

bsutton replied:
done.

mraleph commented on lib/posix.dart:
```diff
@@ -82,8 +111,32 @@ class _PosixConditionVariable extends ConditionVariable {
+
+  /// Waits on a condition variable with a timeout.
+  void _timedWait(Duration timeout, _PosixMutex mutex) {
+    final abstime =
```
The code for computing timespec for waiting should probably be shared between `_timedWait` and `_timedLock`.

bsutton replied:
done.

mraleph commented on lib/src/bindings/pthread.dart:
```diff
@@ -10,29 +10,61 @@ final class pthread_mutex_t extends Opaque {}
+  external int tv_nsec;
+}
+
+/// mutex_init
```
These comments are redundant. They don't add anything.

bsutton replied:
they were added to help visual parsing of the code not to add documentation.

mraleph replied:
Then at the very least make this consistent: 

- in `pthread.dart` you added a comment at every function, in `winapi` you did not. 
- in `pthread.dart` you use function name (minus `pthread_`), in `winapi` you use human readable categories instead.

Also if these are not for documentation then make them `//` rather than `///`.

mraleph commented on lib/src/bindings/pthread.dart:
```diff
@@ -10,29 +10,61 @@ final class pthread_mutex_t extends Opaque {}
 
 final class pthread_cond_t extends Opaque {}
 
+// final class pthread_timespec_t extends Opaque {}
```
Commented code.

bsutton replied:
removed.

mraleph commented on lib/windows.dart:
```diff
@@ -59,13 +61,20 @@ class _WindowsConditionVariable extends ConditionVariable {
+        timeout == null ? infinite : timeout.inMilliseconds, exclusive);
+
+    if (result != 1) {
+      if (GetLastError() == ERROR_TIMEOUT) {
```
Sidenote: unfortunately this is not reliable because `GetLastError` can be clobbered with very low probability when returning from `SleepConditionVariableSRW`. Nothing you can do about it now though.

bsutton replied:
how does it get clobbered? GetLastError is maintained on a per thread basis and there are no other calls after SleepConditionVariableSRW.

mraleph replied:
It can get clobbered because: 

- VM can decide to do a GC when returning from `SleepConditionVariableSRW` at which point it will go to runtime to block the thread (and possibly do GC related actions). 
- `GetLastError` (can be) lazily linked, so when you invoke `GetLastError` for the first time execution would go into runtime to perform resolution and clobber the current value.

mraleph reviewed (changes requested):
Some comments.

bsutton committed:
improved the timeout to support microseconds duration. Fixed a bug in the calculations for remaining time when the wait wakes up for sperious reasons.
Change the while termination in takeTimed to check for _stateEmpty rather than !_stateFull in anticipation of the mailbox supporting multiple messages rather than jus the single message it currently supports.

bsutton committed:
Moved the timepsec creation into a separate function and now support microsecond resolution for timeouts.

bsutton committed:
removed old code.

bsutton committed:
formated with dart format.

bsutton committed:
Merge branch 'main' into main

bsutton committed:
merged in the mailbox.close method from upstream.

bsutton committed:
Merge branch 'main' of github.com:onepub-dev/native_synchronization

bsutton committed:
unit test were still using the temp package name.

bsutton commented:
changes as requested and merged with main.

bsutton committed:
comment out code to help isolate crash

bsutton committed:
commit for dart ffi issue

bsutton commented:
I've just had to raise an issue with the dart team as ffi seems to be crashing with the latest merges from main.

https://github.com/dart-lang/sdk/issues/56412

So perhaps best no to merge this pr until we see a response on the noted issue.

mraleph commented on test/primitives_test.dart:
```diff
@@ -61,18 +71,44 @@ void main() {
+      unawaited(
+          spawnLockedMutex(mutex.asSendable, const Duration(seconds: 10)));
+
+      /// give the isoalte a chance to start.
```
- Typo: `isolate`.
- do not use doc comments inside functions.
- Please capitalize sentences in comments and end them with a `.`.

mraleph replied:
This is still not resolved?

bsutton committed:
changed /// to // as there comments are just to help with visual parsing.

bsutton committed:
added comments in winapi to help break up the code to make it easier to visually parser.

bsutton committed:
changed all print statemetns to use the dart logger and cleaned up the logging messages.
restored static type to finalizer.
restored all of hte unit tests now that we have identified the cause of the crash.

bsutton committed:
Added an initial call to GetLastError for windows to get around the dynamic link of windows dll which can result in other windows system calls between when dart call GetLastError and when it is actually called in the windows subsystem.

bsutton committed:
Released 0.5.0.

bsutton committed:
deployed native_synchronization_temp and then reverted back the package names.

bsutton committed:
restored the call _mailbox.ref.state that cuases the ffi core dump.

bsutton committed:
Merge branch 'ffi-crash'

bsutton committed:
removed the test in _takeTimed for a closed mailbox as a work around for https://github.com/dart-lang/sdk/issues/56412

bsutton commented:
I've removed the check for a closed mailbox from Mailbox.take() with a timeout to side set the FFI crash.

Do we wait for an fix to ffi before we release?

mraleph commented on lib/mailbox.dart:
```diff
@@ -102,10 +104,59 @@ class Mailbox {
+          final remainingTime = _remainingTime(timeout, start);
+          _condVar.wait(_mutex, timeout: remainingTime);
+        }
+        // if (_mailbox.ref.state == _stateClosed) {
```
Commented out code?

mraleph commented:
Tests are not passing and there are still unresolved comments.

bsutton committed:
Added method to the public API to determine the mailboxes state.
Added so a user can determine what a call to Mailbox.put failed (e.g. the box was full or closed).

bsutton committed:
Released 0.6.0.

bsutton committed:
Applied dart format to code.

devoncarew commented:
Hi - thanks for the PR. FYI that the source-of-truth for this package has moved to our dart-lang/labs monorepo; https://github.com/dart-lang/labs/tree/main/pkgs/native_synchronization.

I'll close this PR, but feel free to re-create it against the new package location if this PR is still active.

mraleph commented:
The PR was stalled mostly because I was distracted - I will take a landing this code in the monorepo when I get a chance.
